### ISSUE TITLE
ENH: optmize.slsqp: return Lagrange multipliers

### ISSUE THREAD

**Issue by jaharris85** [scipy/scipy#14394]
Hello,

I was wondering whether anyone had tested the accuracy of the Lagrange multipliers returned by the minimizer method SLSQP? 

I created a very simple problem of minimizing (x - c)^2 with bounds 1 <= x <= 2. The solution returned by the solver for x is accurate to within the tolerance set. The multipliers I have plotted in the attachment and show a consistent and noticeable offset. I know that the multipliers are not officially returned by the minimizer method but I would expect that accurate multipliers are needed to ensure that the gradient of the Lagrangian is zero at the solution?

Thanks.

![multipliers](https://user-images.githubusercontent.com/86674313/125315276-db2b9780-e32e-11eb-98eb-5e358caae691.png)

#### Reproducing code example:

```
import numpy as np
import scipy.optimize as opt
from matplotlib import pyplot as plt

ineq_cons = {'type' : 'ineq',
             'fun' : lambda x: np.array([x[0] - 1.0, 2.0 - x[0]]),
             'jac' : lambda x: np.array([[1.0], [-1.0]])}

c = np.linspace(0, 3, num = 101)

x0 = np.array([1.5])

x = np.zeros(len(c))
m = np.zeros(len(c))
l = np.zeros((2, len(c)))

for i in range(len(c)):
    fun = lambda x: np.array([(x[0] - c[i]) * (x[0] - c[i])])
    jac = lambda x: np.array([2 * (x[0] - c[i])])
    
    res, w = opt.minimize(fun, x0, method = 'SLSQP', jac = jac, constraints = ineq_cons)
    
    x[i] = res.x
    l[0, i] = w[0]
    l[1, i] = w[1]

    if c[i] < 1:
        m[i] = 2 - 2 * c[i]
    elif c[i] > 2:
        m[i] = 2 * c[i] - 4
    else:
        m[i] = 0.0

plt.title("Min (x - c)^2 s.t. 1 <= x <= 2")
plt.xlabel("c")
plt.ylabel("lambda(c)")
plt.plot(c, l[0,:], 's', label="x >= 1")
plt.plot(c, l[1,:], 'o', label="x <= 2")
plt.plot(c, m, 'k', label="Analytical")
plt.grid()
plt.legend(loc="upper center")
plt.show()
```

#### Scipy/Numpy/Python version information:
```
1.7.0 1.21.0 sys.version_info(major=3, minor=7, micro=7, releaselevel='final', serial=0)
```

**Comment by andyfaff** [comment]
This seems like more of a usage question. Github issues are intending mainly for bugs, rather than usage. It's probably better to direct usage questions to the scipy mailing list.

The example as written doesn't work because `minimize` returns an `OptimizeResult`, which isn't something that can be unpacked into two variables. You can find out what SLSQP returns [here](https://github.com/scipy/scipy/blob/master/scipy/optimize/slsqp.py#L460). That being the case I don't know what you've actually plotted in the graph, or how it was obtained.

For this case you're better off making bounds for the parameter,`bounds=[(1, 2)]`, and providing them as the `bounds` keyword to `minimize`, it's much simpler than constraints for this example.

**Comment by jaharris85** [comment]
Sorry, I forgot to mention that I had to modify slsqp.py to return the first m entries of w as the second return argument:

```
    return OptimizeResult(x=x, fun=fx, jac=g[:-1], nit=int(majiter),
                          nfev=sf.nfev, njev=sf.ngev, status=int(mode),
                          message=exit_modes[int(mode)], success=(mode == 0)), w[:m]
```

Reading the comments in slsqp_optmz.f, the variable w should contain the Lagrange multipliers once the solver has converged.

You're right that it would be a lot easier just to use ```bounds```, but I needed to create a simple enough example that I could solve by hand to test the accuracy of the Lagrange multipliers; the multipliers are incredibly useful for sensitivity analysis in more complicated problems and I wanted to check they were correct. Because they seem so inaccurate it also makes me wonder whether SLSQP is even converging correctly to the true minimum.

**Comment by paulduf** [comment]
Hello, I'm running into the same issue (I need to get the lagrange multipliers from the slsqp routine to identify active constraints in the KKT equation). There is also [this old mail sequence](https://mail.python.org/pipermail/scipy-user/2010-March/024487.html) where Guilherme P. de Freitas already noticed the Lagrange multipliers are not very accurate. I don't really know what these multipliers really refer to, I guess they are the multipliers of the approximate, quadratic subproblem. But you proposed a linearly constrained quadratic problem as a test example here and it's still inaccurate, so I'm a bit confused. I would guess slsqp converges before the BFGS update is able to properly estimate the true Hessian of the problem.

Anyway for my use case I'm fine with these approximate multipliers, as a technical remark I would add that you don't break the current interface by adding the multipliers to the results dict, like this:

```python
 return OptimizeResult(x=x, fun=fx, jac=g[:-1], lag=w[:m], nit=int(majiter),
                          nfev=sf.nfev, njev=sf.ngev, status=int(mode),
                          message=exit_modes[int(mode)], success=(mode == 0))
```

Since this feature has been already asked more than 11 years ago, is there a reason it is not available in the scipy interface ?

**Comment by andyfaff** [comment]
> Since this feature has been already asked more than 11 years ago, is there a reason it is not available in the scipy interface ?

The project welcomes pull requests to add such functionality. Open source projects are always on the lookout for additional skilled people to effect change. For my part, although I contribute to `optimize` a lot, I don't have a grounding in the maths behind Lagrange multipliers + KKT conditions, so it's harder for me to make that change. If you have that knowledge it would be great to receive a PR.

Whilst it's relatively easy to return those multipliers there are a few things to think about:

1. There needs to be adequate documentation of what the Lagrangian multiplier output is.
2. de Freitas noted in the [original mailing list discussion](https://mail.python.org/pipermail/scipy-user/2010-March/024608.html):

> I was wondering if it is worth outputting the multipliers at all... The low accuracy in a very simple example (see my previous post) makes me wonder how innacurate the estimated multipliers can be in a more complex problem.

Perhaps there needs to be a quantification of this before we add them to OptimizeResult? If the Lagrange multipliers are being used to figure out active constraints then presumably a certain level of accuracy is required in them. 
If there is room for improvement in the Lagrange multiplier accuracy, then as well as understanding the maths, the PR would require knowledge of the SLSQP Fortran codebase.

If you wish to estimate if constraints are active, then is it possible to use the constraint functions and `OptimizeResult.x` to see how close the constraints are to being violated?

**Comment by paulduf** [comment]
I agree with you and I plan to further inspect the Lagrange multipliers returned by the Fortran code. Let me just clarify this:

> If you wish to estimate if constraints are active, then is it possible to use the constraint functions and OptimizeResult.x to see how close the constraints are to being violated?

First, this requires the user to define a threshold under which a constraint is considered active. Moreover, using the Lagrange multipliers, we can obtain the _strictly_ active set at `x`, that is the set of constraints associated with a strictly positive Lagrange multiplier, which is stronger than the constraint being active. Imagine your problem violates the [LICQ](https://en.wikipedia.org/wiki/Karush%E2%80%93Kuhn%E2%80%93Tucker_conditions#Regularity_conditions_(or_constraint_qualifications)), then you can have an active constraint (so `g(x) == 0`) but the Lagrange multiplier is still 0, because there exists a linear combination of other constraints gradients at `x` associated with strictly positive multipliers.

An additional remark, as I've been using SLSQP a lot these days, people may request other features of this kind. I would like to output the BFGS estimate of the Hessian matrix as well, for further analysis of my optimization problem (haven't tried this so far).

**Comment by paulduf** [comment]
I just found out that this issue is a duplicate of this  one: #9827

And, if this vector is exposed, it must be clear that the vector is *an estimate of the Lagrange multipliers* and not the exact Lagrange multipliers, because SLSQP is a quasi-Newton method and these multipliers are estimated from the BFGS matrix. Moreover, I guess because it is an active set method, the solver candidate solution converges to the constrained optimum faster than the BFGS matrix converges to the true Hessian, hence the low accuracy in the Lagrange multipliers' estimates.

But the good news is that whether the Lagrange multiplier is strictly positive or zero seems not to suffer from this low accuracy, so we can reliably use this fact in higher-level optimization code using SLSQP as a subroutine.

So here is reproducible python code with a few experiments to support my claims. 
- from the first two, you optimize a sphere function (sum of squares) under 3 orthogonal linear constraints in the first case, which are all active, and the true Lagrange multipliers are [2, 2, 2]. In the second case, the first constraint is strictly active (aligned with the gradient), hence the second is not need to obtain the KKT stationarity condition. The value for the Lagrange estimate is very close to 0.
- from the second two, you optimize the same problem, and ellipsoid function (like a sphere with non-identity hessian matrix) under the 3 axis-aligned constraints, but starting from a different point (either [10, 10, 10] or [100, 100, 100]) and you clearly see the influence of this fact on the multipliers estimates.

```python
from scipy.optimize import fmin_slsqp

def objective(x):
    return sum(a * xi**2 for a, xi in zip(diag, x))

def constraint(x):
    return [xi - 1 for xi in x]

def constraint2(x):
    return [sum(x) - len(x), x[0] - 1]

diag = np.ones(3)
print(" >>> Sphere function, 3 strictly active constraints")
res = fmin_slsqp(objective, np.ones(3) * 10, f_ieqcons=constraint, full_output=True)
print(res.lag)

print(" >>> Sphere function, 1 strictly active and 1 active constraint")
res = fmin_slsqp(objective, np.ones(3) * 10, f_ieqcons=constraint2, full_output=True)
print(res.lag)

diag = [10, 5, 2]
print(" >>> Ellipsoid function, 3 strictly active constraints")
res = fmin_slsqp(objective, np.ones(3) * 10, f_ieqcons=constraint, full_output=True)
print(res.lag)

diag = [10, 5, 2]
print(" >>> Ellipsoid function, 3 strictly active constraints, starting point further away")
res = fmin_slsqp(objective, np.ones(3) * 100, f_ieqcons=constraint, full_output=True)
print(res.lag)
```

And the output on my computer:
```console
>>> Sphere function, 3 strictly active constraints
Optimization terminated successfully    (Exit mode 0)
            Current function value: 2.9999999999948876
            Iterations: 2
            Function evaluations: 8
            Gradient evaluations: 2
[6.5 6.5 6.5]
 >>> Sphere function, 1 strictly active and 1 active constraint
Optimization terminated successfully    (Exit mode 0)
            Current function value: 3.000000000005585
            Iterations: 2
            Function evaluations: 8
            Gradient evaluations: 2
[6.50000000e+00 1.31469857e-14]
 >>> Ellipsoid function, 3 strictly active constraints
Optimization terminated successfully    (Exit mode 0)
            Current function value: 17.000000025410714
            Iterations: 2
            Function evaluations: 8
            Gradient evaluations: 2
[105.49999999  50.5         17.5       ]
 >>> Ellipsoid function, 3 strictly active constraints, starting point further away
Optimization terminated successfully    (Exit mode 0)
            Current function value: 17.000000000000156
            Iterations: 3
            Function evaluations: 12
            Gradient evaluations: 3
[490.25000432 232.74999835  78.24999861]
```

**Comment by rgommers** [comment]
> I just found out that this issue is a duplicate of this one: #9827

Thanks @paulduf, this is a good observation. This issue has more information, but gh-9827 already has a linked PR and is older. So I will close this issue and cross-link back here from gh-9827.

**Comment by andyfaff** [comment]
@rgommers, this issue was a two parter:

1. Why doesn't SLSQP return the multipliers
2. Why are the multipliers that are returned inaccurate.

#9827 addresses the first part and there are two PR attempts to closing that issue (#15641, #9839). However, neither of these address the accuracy part. I'd like to reopen this issue whilst the accuracy is investigated.

**Comment by rgommers** [comment]
Thanks @andyfaff, sorry about that.

**Comment by andyfaff** [comment]
The PR in #15641 returns the KKT multipliers, as calculated by SLSQP.
They are much closer to the analytical plot in your first graph.
![PR](https://user-images.githubusercontent.com/702934/155422281-02e3f502-0cd4-4c16-9e75-1c3ce13b1322.png)

@paulduf I'm going to add the first example as a test, but would like to add the sphere and ellipsoidal examples you provided on 6/Dec as extra tests. Can you tell me what you expect the KKT multipliers to be for all four examples?

**Comment by paulduf** [comment]
It's not easy to write math on GitHub, so I attached a small pdf to explain the computation. I haven't proof-read it, so please tell me if you find some errors.

[scipy_slsqp_kkt.pdf](https://github.com/scipy/scipy/files/8131501/scipy_slsqp_kkt.pdf)

**Comment by ilayn** [comment]
The multiplier estimates are now included in the new version #22524. Similar to #15641, now the approximations come in the Result object field `multipliers` 

Hence this line 

```Python
     res, w = opt.minimize(fun, x0, method = 'SLSQP', jac = jac, constraints = ineq_cons)
```

now reads as 

```Python
     res = opt.minimize(fun, x0, method = 'SLSQP', jac = jac, constraints = ineq_cons)
     w = res['multipliers']
```

and the results are similar, 


![Image](https://github.com/user-attachments/assets/cb77984f-54f2-424c-9236-83cfc751ad5d)

**Comment by ilayn** [comment]
@andyfaff If you want to add the variable bound multipliers on top of what SLSQP returns (only constraint multipliers are considered for now), we can carry that over from #15641

**Comment by mdhaber** [comment]
@ilayn just thought I'd mention that the tests from gh-15641, which check that the multipliers are consistent with the KKT conditions, should still be valuable. Please include them in gh-22524.

**Comment by ilayn** [comment]
@mdhaber I want to know if we are returning multipliers corresponding to variable bounds. Otherwise we test SLSQP code which is what it is for the last 35 years.

**Comment by ilayn** [comment]
The multipliers corresponding to the problem constraints are returning as is, we are not touching them. So not sure what the test should be then other than the variables are returned. That's exactly the same point I made during our lapack discussion that we don't need to test array slicing.

On the other hand, if we need to add the bound multipliers that's computation on our side so that's what we need to test.

**Comment by ilayn** [comment]
OK, that makes sense. Then should I get also the bound multipliers from #15641 ? Not sure if I understood what the PR is doing but I can try to adapt.

**Comment by mdhaber** [comment]
Yes, please. I don't know much about how the bound multipliers were processed, so I can't help there, but once that's ready, I can update the tests and submit a PR with those to your branch.

**Comment by ilayn** [comment]
I actually looked at the code but I don't think those tests are doing what they are supposed to be doing. Because the order of the multipliers coming from Fortran are not returned as the tests assume they are. This would need a bit more surgery I'm afraid to sort the finitely bounded variables and should check if there is any and so on.

Essential difficulty is the following, say you have 3 eq and 2 ineq constraints and you bounded only x3 and x5 from below, and x1, and x4 from above. The multipliers must be returned as 

```python
mult = [EQ, EQ, EQ, INEQ, INEQ, LOW3, LOW5, UPP1, UPP4]
```

And you have to go through the bound constraints to match the finite ones with these multipliers. But this assumption is not correct #15641 because Fortran code returns 

```python
mult = [EQ, EQ, EQ, INEQ, INEQ, LOW3, LOW5, JUNK, UPP1, UPP4, JUNK, JUNK]
```

or 

```python
mult = [EQ, EQ, EQ, INEQ, INEQ, LOW3, LOW5, UPP1, UPP4, JUNK, JUNK, JUNK]
```

This is because there is an augmentation happening in case the linearized problem is not solved though original problem is still solvable. Hence this one here cannot be investigated from outside without being cleaned from the low-level side. It's not impossible but needs some surgery.  I propose we do it as a follow-up after #22524  because it is already getting a bit too crowded in there.

**Comment by mdhaber** [comment]
I am fine with multipliers waiting for a follow-up PR, but:

> I actually looked at the code but I don't think those tests are doing what they are supposed to be doing.

This is phrased as though the tests were written incorrectly. Is that what you mean?

These tests were written for a PR in which there was intermediate Python code that was supposed to take the multipliers returned by the Fortran and produce a result dictionary satisfying a documented interface. 

If you mean that the tests were not passing in *that* PR, was it because the Python intermediary code was not doing its job to satisfy the documented interface, or was the test function incorrect in its interpretation of the interface? Or do you mean that even if the existing tests passed in that PR, the generic, propery-based tests would not pass for _more general_ test problems because the Python intermediary code would not always produce the results according to the documented interface?

Hopefully this makes it clear why tests are needed, assuming we are going to document the interface. But currently it looks like gh-22524 just returns the raw multipliers as produced by the Fortran code in an undocumented `multipliers` attribute. Is the plan really to return them like that - i.e., as undocumented "advanced features" that are there if the user can find them, and it's up to the user to figure out how they correspond with the input?

**Comment by ilayn** [comment]
The bounds multipliers are currently NaN'd in the fortran code. So whatever was being sliced would have been NaN regardless. 

> Hopefully this makes it clear why tests are needed, assuming we are going to document the interface

We are not documenting the interface in case of bound multipliers. This is an ENH that was not possible before. Fortran code only returns the equality and inequality constraint multipliers and the rest is discarded inside the F77 code. We are trying to recreate it.

The main hurdle is matching the finite bounds on the variables which is given as `(NaN, NaN,5,10,NaN...)` for `xl` and (NaN, 3,NaN,6,NaN...) etc for `xu` to the corresponding multipliers which comes back as `(num, num, num, ...)` with no structure. Also within those `num`s two of them might be the internal augmented variable that the low-level code needs to take out first before sending. So it is not that trivial to match the Py side to the C side. Hence the prior PR could not do this without modifying Fortran code.

We can recreate the multipliers directly from the solution anyways by solving a least squares problem and marking the others as zero but then we have to find the active constraints but either way it is some surgery for a new feature and not a property of SLSQP.

**Comment by ev-br** [comment]
Bumping the milestone to match gh-22787
